File: src/services/egress-identity.ts
import { invoke } from '@tauri-apps/api/core'

export interface EgressPreviewRequest {
  process_name?: string
  exe_path?: string
  shortcut_id?: string
  domain?: string
  source_ip?: string
  source_port?: number
  available_nodes?: string[]
}

export interface ResolvedEgressIdentity {
  assignmentKey?: string | null
  profileId: string
  selectedNode: string
  dnsMode: 'Inherit' | 'Hijack' | 'Remote'
  tlsFingerprint?: string | null
  matchedBy: string
}

export async function egressIdentityPreviewMatch(
  request: EgressPreviewRequest,
): Promise<ResolvedEgressIdentity> {
  return await invoke('egress_identity_preview_match', {
    ...request,
  })
}

export async function egressIdentityAssignMatch(
  request: EgressPreviewRequest,
): Promise<ResolvedEgressIdentity> {
  return await invoke('egress_identity_assign_match', {
    ...request,
  })
}

export async function egressIdentityGetActiveAssignments(): Promise<ResolvedEgressIdentity[]> {
  return await invoke('egress_identity_get_active_assignments')
}

export async function egressIdentityClearAssignment(key: string): Promise<void> {
  await invoke('egress_identity_clear_assignment', { key })
}

File: src-tauri/src/cmd/egress_identity.rs
use crate::core::egress_identity::{EgressIdentityConfig, ResolvedEgressIdentity};

use super::{CmdResult, StringifyErr};

#[tauri::command]
pub fn egress_identity_get_config() -> CmdResult<EgressIdentityConfig> {
    Ok(crate::feat::egress_identity_get_config())
}

#[tauri::command]
pub async fn egress_identity_preview_match(
    process_name: Option<String>,
    exe_path: Option<String>,
    shortcut_id: Option<String>,
    domain: Option<String>,
    source_ip: Option<String>,
    source_port: Option<u16>,
    available_nodes: Option<Vec<String>>,
) -> CmdResult<ResolvedEgressIdentity> {
    crate::feat::egress_identity_preview_match(
        process_name, exe_path, shortcut_id, domain, source_ip, source_port, available_nodes,
    )
    .await
    .stringify_err()
}

#[tauri::command]
pub async fn egress_identity_assign_match(
    process_name: Option<String>,
    exe_path: Option<String>,
    shortcut_id: Option<String>,
    domain: Option<String>,
    source_ip: Option<String>,
    source_port: Option<u16>,
    available_nodes: Option<Vec<String>>,
) -> CmdResult<ResolvedEgressIdentity> {
    crate::feat::egress_identity_assign_match(
        process_name, exe_path, shortcut_id, domain, source_ip, source_port, available_nodes,
    )
    .await
    .stringify_err()
}

#[tauri::command]
pub fn egress_identity_get_active_assignments() -> CmdResult<Vec<ResolvedEgressIdentity>> {
    Ok(crate::feat::egress_identity_get_active_assignments())
}

#[tauri::command]
pub fn egress_identity_clear_assignment(key: String) -> CmdResult<()> {
    crate::feat::egress_identity_clear_assignment(&key);
    Ok(())
}
